From c0c96214ec46a30997fe1dcf15bbd7802c982816 Mon Sep 17 00:00:00 2001
From: Nathan Muggli <nmuggli@google.com>
Date: Thu, 7 May 2026 11:25:20 -0600
Subject: [PATCH] disable boringssl in base::RandBytes

Introduced in r1045239 (crrev.com/c/3880863).
Updated in r1321254 (crrev.com/c/5669584).
Updated in r1477790 (crrev.com/c/6665689).
Updated in r1624070 (crrev.com/c/7812144).
Updated in r1626560 (crrev.com/c/7828277).

In the most recent update, upstream removed all of the previous
functionality this patch has been using, so this adds it back in.  Also,
this was combined with the long-term-0900 patch since they are both
related.
---
 base/rand_util_posix.cc | 42 +++++++++++++++++++++++++++++++++++++----
 1 file changed, 38 insertions(+), 4 deletions(-)

diff --git a/base/rand_util_posix.cc b/base/rand_util_posix.cc
index b37f87159e..099105a4e2 100644
--- a/base/rand_util_posix.cc
+++ b/base/rand_util_posix.cc
@@ -12,9 +12,8 @@
 #include "base/posix/eintr_wrapper.h"
 #include "base/system/sys_info.h"
 #include "build/build_config.h"
-#include "third_party/boringssl/src/include/openssl/rand.h"
 
-#if BUILDFLAG(IS_MAC)
+#if BUILDFLAG(IS_MAC) || 1
 // TODO(crbug.com/40641285): Waiting for this header to appear in the iOS SDK.
 // (See below.)
 #include <sys/random.h>
@@ -42,6 +41,27 @@ double RandomUint64ToRandomDouble(uint64_t value) {
   return (value >> 11) * 0x1.0p-53;
 }
 
+bool KernelSupportsGetRandom() {
+  return base::SysInfo::KernelVersionNumber::Current() >=
+         base::SysInfo::KernelVersionNumber(3, 17);
+}
+
+bool GetRandomSyscall(void* output, size_t output_length) {
+  // We have to call `getrandom` via Linux Syscall Support, rather than through
+  // the libc wrapper, because we might not have an up-to-date libc (e.g. on
+  // some bots).
+  const ssize_t r =
+      HANDLE_EINTR(syscall(__NR_getrandom, output, output_length, 0));
+
+  // Return success only on total success. In case errno == ENOSYS (or any other
+  // error), we'll fall through to reading from urandom below.
+  if (output_length == static_cast<size_t>(r)) {
+    MSAN_UNPOISON(output, output_length);
+    return true;
+  }
+  return false;
+}
+
 }  // namespace
 
 namespace internal {
@@ -66,8 +86,22 @@ double RandDoubleAvoidAllocation() {
 }  // namespace internal
 
 void RandBytes(span<uint8_t> output) {
-  // BoringSSL's RAND_bytes always returns 1. Any error aborts the program.
-  (void)RAND_bytes(output.data(), output.size());
+  // On Android it is mandatory to check that the kernel _version_ has the
+  // support for a syscall before calling. The same check is made on Linux and
+  // ChromeOS to avoid making a syscall that predictably returns ENOSYS.
+  static const bool kernel_has_support = KernelSupportsGetRandom();
+  if (kernel_has_support && GetRandomSyscall(output.data(), output.size())) {
+    return;
+  }
+
+  // If the OS-specific mechanisms didn't work, fall through to reading from
+  // urandom.
+  //
+  // TODO(crbug.com/40641285): When we no longer need to support old Linux
+  // kernels, we can get rid of this /dev/urandom branch altogether.
+  const int urandom_fd = GetUrandomFD();
+  const bool success = ReadFromFD(urandom_fd, as_writable_chars(output));
+  CHECK(success);
 }
 
 int GetUrandomFD() {
-- 
2.54.0.563.g4f69b47b94-goog

